pkg/queue/breaker.go

```diff
@@ -170,52 +163,42 @@ func (s *Semaphore) Release() error {
 	}
 }
 
-// AddCapacity increases the number of tokens in the rotation.
-// Increases only if the size is positive, does nothing otherwise.
-// An error is returned if not all capacity could be added to the semaphore.
-// This error could happen when the number of tokens exceeds the semaphore's queue size.
-func (s *Semaphore) AddCapacity(size int32) error {
+// UpdateCapacity updates the capacity of the semaphore to the desired
+// size.
+func (s *Semaphore) UpdateCapacity(size int32) error {
 	s.mux.Lock()
 	defer s.mux.Unlock()
 
-	for i := int32(0); i < size; i++ {
-		select {
-		case s.queue <- s.token:
-			s.capacity++
-		default:
-			return ErrAddCapacity
-		}
+	// Effective capacity is the capacity after taking reducers
+	// into account.
+	effectiveCapacity := s.capacity - s.reducers
+
+	if effectiveCapacity == size {
+		return nil
 	}
-	return nil
-}
 
-// ReduceCapacity removes tokens from the rotation.
-// It tries to acquire as many tokens as possible. If there are not enough tokens in the queue,
-// it postpones the operation for the future by increasing the `reducers` counter.
-// We return an error when attempting to reduce more capacity than was originally added.
-func (s *Semaphore) ReduceCapacity(size int32) error {
-	s.mux.Lock()
-	defer s.mux.Unlock()
+	if size > s.maxCapacity {
+		return ErrAddCapacity
+	}
 
-	if size+s.reducers > s.capacity {
+	if size < 0 {
 		return ErrReduceCapacity
 	}
 
-	for i := int32(0); i < size; i++ {
-		select {
-		case <-s.queue:
-			s.capacity--
-		default:
-			s.reducers++
+	if effectiveCapacity < size {
+		for ; s.capacity-s.reducers < size; s.capacity++ {
```

Comment:
Why don't you first reduce reducers?

Reply:
I will, that's why it's still WIP. It was too late in the day yesterday to accurately make these decisions. Fresh mind, fresh look now! 🙂